### ISSUE TITLE
Memory leak in stats.rv_continuous.ppf on MacOS Sonoma

### ISSUE THREAD

**Issue by djgagne** [scipy/scipy#20475]
I have been working on a project to perform a quantile transform followed by norm.ppf to retrieve the z-score values for each quantile on a multi-TB dataset in parallel. However, I keep running into out of memory issues when doing so. After running memory_profiler on my code, one of the sources of memory leaks appears to be coming from my call to norm.ppf. When I run norm.ppf, the memory increment is much larger than the input data array. When I run the underlying ndtri function by itself, there is no memory increment as one would expect.  
```
   119    561.5 MiB     19.1 MiB           1       output_arr = np.full((1000, 50, 50), 0.5)
   120    700.2 MiB    138.6 MiB           1       output_arr = norm.ppf(output_arr)
   121    700.2 MiB      0.0 MiB           1       output_arr = np.full((1000, 50, 50), 0.5)
   122    700.2 MiB      0.0 MiB           1       output_arr = ndtri(output_arr)
```
My best guess is that there is some kind of memory issue going on in the following code block, but I am not sure where it is coming from. Anyone have an idea?

https://github.com/scipy/scipy/blob/7dcd8c59933524986923cde8e9126f5fc2e6b30b/scipy/stats/_distn_infrastructure.py#L2214-L2257

**Comment by mdhaber** [comment]
A leak is different from temporarily consuming more memory than expected. All distributions consume much more memory than the underlying special functions due to various input validation and output correction steps. Does this extra memory usage persist after you finish the function, or is it temporary?

**Comment by djgagne** [comment]
The memory usage persists well after the ppf function is called, including when the code transitions to a different scope and any intermediate arrays should have been garbage collected.

**Comment by mdhaber** [comment]
OK, can you provide a MRE with `psutil.Process().memory_info().rss` or another tool to measure memory usage?

**Comment by djgagne** [comment]
Here is a MRE. It looks like ndtri also does not give up memory but at a much slower growth rate than norm.ppf. Both have issues even with manual gc.collect().
```python
from scipy.stats import norm
from scipy.special import ndtri
import numpy as np
import matplotlib.pyplot as plt
import psutil
import gc

process = psutil.Process()
n_elements = 301
mem_vals = np.zeros(n_elements)
mem_vals[0] = process.memory_info().rss / 1e6
for i in range(1, n_elements):
    x = np.random.random(size=(100, 50, 50))
    ppf_val = ndtri(x)
    mem_vals[i] = process.memory_info().rss / 1e6
    gc.collect()
plt.plot(mem_vals[1:] - mem_vals[0], label="ndtri")
mem_vals = np.zeros(n_elements)
mem_vals[0] = process.memory_info().rss / 1e6
for i in range(1, n_elements):
    x = np.random.random(size=(100, 50, 50))
    ppf_val = norm.ppf(x)
    mem_vals[i] = process.memory_info().rss / 1e6
    gc.collect()
plt.plot(mem_vals[1:] - mem_vals[0], label="norm.ppf")
plt.xlabel("Iterations")
plt.ylabel("Memory usage (MB)")
plt.legend()
plt.savefig("norm_usage_tracking.png", dpi=200, bbox_inches="tight")
```
![norm_usage_tracking](https://github.com/scipy/scipy/assets/874553/8755d827-8fdd-4ef3-98b4-2ec8445f3f9e)

**Comment by djgagne** [comment]
I have also added a control line where I do the same loop but without calling either norm.ppf or ndtri. 
![norm_usage_tracking](https://github.com/scipy/scipy/assets/874553/e3aa4a61-f170-4bdb-9bb1-55db0132b5fe)

**Comment by j-bowhay** [comment]
I cannot reproduce on main, what version of SciPy are you using?

**Comment by djgagne** [comment]
I am using scipy 1.12.0 and numpy 1.23.5 on MacOS Sonoma 14.3.1. Here is my full conda environment for reference. I have also seen similar issues on SUSE enterprise Linux. 
```
# packages in environment at /Users/dgagne/miniconda3/envs/bridgescaler:
#
# Name                    Version                   Build  Channel
_ipython_minor_entry_point 8.7.0                hb6b4a82_0    conda-forge
accessible-pygments       0.0.4                    pypi_0    pypi
alabaster                 0.7.16             pyhd8ed1ab_0    conda-forge
appnope                   0.1.3              pyhd8ed1ab_0    conda-forge
asttokens                 2.2.0              pyhd8ed1ab_0    conda-forge
attrs                     22.1.0             pyh71513ae_1    conda-forge
babel                     2.14.0             pyhd8ed1ab_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                pyhd8ed1ab_3    conda-forge
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
beautifulsoup4            4.12.3             pyha770c72_0    conda-forge
bleach                    5.0.1                    pypi_0    pypi
bridgescaler              0.6.0                     dev_0    <develop>
brotli                    1.0.9                hb7f2c08_8    conda-forge
brotli-bin                1.0.9                hb7f2c08_8    conda-forge
brotli-python             1.0.9           py310h7a76584_8    conda-forge
build                     0.9.0                    pypi_0    pypi
bzip2                     1.0.8                h0d85af4_4    conda-forge
c-ares                    1.18.1               h0d85af4_0    conda-forge
ca-certificates           2024.2.2             h8857fd0_0    conda-forge
certifi                   2022.9.24                pypi_0    pypi
cftime                    1.6.2           py310h936d966_1    conda-forge
charset-normalizer        2.1.1                    pypi_0    pypi
clean                     0.1.4                    pypi_0    pypi
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
commonmark                0.9.1                    pypi_0    pypi
conda-souschef            2.2.3              pyhd8ed1ab_0    conda-forge
contourpy                 1.0.7           py310ha23aa8a_0    conda-forge
crick                     0+untagged.55.g8ec0b07.dirty          pypi_0    pypi
curl                      7.86.0               h581aaea_1    conda-forge
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
docutils                  0.19                     pypi_0    pypi
exceptiongroup            1.0.4              pyhd8ed1ab_0    conda-forge
executing                 1.2.0              pyhd8ed1ab_0    conda-forge
fonttools                 4.39.4          py310h6729b98_0    conda-forge
freetype                  2.12.1               h3f81eb7_1    conda-forge
grayskull                 2.5.3              pyhd8ed1ab_0    conda-forge
hdf4                      4.2.15               h7aa5921_5    conda-forge
hdf5                      1.12.2          nompi_h1f71328_100    conda-forge
icu                       70.1                 h96cf925_0    conda-forge
idna                      3.4                      pypi_0    pypi
imagesize                 1.4.1              pyhd8ed1ab_0    conda-forge
importlib-metadata        5.1.0                    pypi_0    pypi
iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge
ipython                   8.7.0              pyhd1c38e8_0    conda-forge
jaraco-classes            3.2.3                    pypi_0    pypi
jedi                      0.18.2             pyhd8ed1ab_0    conda-forge
jinja2                    3.1.3              pyhd8ed1ab_0    conda-forge
joblib                    1.2.0              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   hac89ed1_2    conda-forge
keyring                   23.11.0                  pypi_0    pypi
kiwisolver                1.4.4           py310ha23aa8a_1    conda-forge
krb5                      1.19.3               hb98e516_0    conda-forge
lcms2                     2.15                 h29502cd_0    conda-forge
lerc                      4.0.0                hb486fe8_0    conda-forge
libblas                   3.9.0           16_osx64_openblas    conda-forge
libbrotlicommon           1.0.9                hb7f2c08_8    conda-forge
libbrotlidec              1.0.9                hb7f2c08_8    conda-forge
libbrotlienc              1.0.9                hb7f2c08_8    conda-forge
libcblas                  3.9.0           16_osx64_openblas    conda-forge
libcurl                   7.86.0               h581aaea_1    conda-forge
libcxx                    16.0.6               hd57cbcb_0    conda-forge
libdeflate                1.17                 hac1461d_0    conda-forge
libedit                   3.1.20191231         h0678c8f_2    conda-forge
libev                     4.33                 haf1e3a3_1    conda-forge
libffi                    3.4.2                h0d85af4_5    conda-forge
libgfortran               5.0.0           13_2_0_h97931a8_3    conda-forge
libgfortran5              13.2.0               h2873a65_3    conda-forge
libiconv                  1.17                 hac89ed1_0    conda-forge
liblapack                 3.9.0           16_osx64_openblas    conda-forge
libllvm14                 14.0.6               hc8e404f_4    conda-forge
libnetcdf                 4.8.1           nompi_hc61b76e_106    conda-forge
libnghttp2                1.47.0               h5aae05b_1    conda-forge
libopenblas               0.3.21          openmp_h429af6e_3    conda-forge
libpng                    1.6.39               ha978bb4_0    conda-forge
libsqlite                 3.40.0               ha978bb4_0    conda-forge
libssh2                   1.10.0               h47af595_3    conda-forge
libtiff                   4.5.0                hee9004a_2    conda-forge
libwebp-base              1.3.0                hb7f2c08_0    conda-forge
libxcb                    1.13              h0d85af4_1004    conda-forge
libxml2                   2.10.3               hb9e07b5_0    conda-forge
libzip                    1.9.2                h6db710c_1    conda-forge
libzlib                   1.2.13               hfd90126_4    conda-forge
line-profiler             4.1.2                    pypi_0    pypi
llvm-openmp               18.1.2               hb6ac08f_0    conda-forge
llvmlite                  0.42.0          py310h7d48a1f_1    conda-forge
markupsafe                2.1.5           py310hb372a2b_0    conda-forge
matplotlib                3.7.1           py310h2ec42d9_0    conda-forge
matplotlib-base           3.7.1           py310he725631_0    conda-forge
matplotlib-inline         0.1.6              pyhd8ed1ab_0    conda-forge
memory-profiler           0.61.0                   pypi_0    pypi
more-itertools            9.0.0                    pypi_0    pypi
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
ncurses                   6.3                  h96cf925_1    conda-forge
netcdf4                   1.6.2           nompi_py310h6892ea4_100    conda-forge
numba                     0.59.1          py310h1d5af72_0    conda-forge
numba-stats               1.7.0                    pypi_0    pypi
numpy                     1.23.5          py310h1b7c290_0    conda-forge
openjpeg                  2.5.0                h13ac156_2    conda-forge
openssl                   3.2.1                hd75f5a5_1    conda-forge
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pandas                    1.5.3           py310hecf8f37_1    conda-forge
parso                     0.8.3              pyhd8ed1ab_0    conda-forge
pep517                    0.13.0                   pypi_0    pypi
pexpect                   4.8.0              pyh1a96a4e_2    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    9.4.0           py310h306a057_1    conda-forge
pip                       22.3.1             pyhd8ed1ab_0    conda-forge
pkginfo                   1.9.2                    pypi_0    pypi
pluggy                    1.0.0           py310h2ec42d9_4    conda-forge
progressbar2              4.3.2              pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.33             pyha770c72_0    conda-forge
psutil                    5.9.8                    pypi_0    pypi
pthread-stubs             0.4               hc929b4f_1001    conda-forge
ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge
pure_eval                 0.2.2              pyhd8ed1ab_0    conda-forge
pydata-sphinx-theme       0.15.2                   pypi_0    pypi
pygments                  2.17.2             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1           py310h2ec42d9_5    conda-forge
pytdigest                 0.1.5                    pypi_0    pypi
pytest                    7.2.0           py310h2ec42d9_1    conda-forge
python                    3.10.8          he7542f4_0_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-utils              3.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.10                    3_cp310    conda-forge
pytz                      2022.6             pyhd8ed1ab_0    conda-forge
pyudorandom               1.0.0                    pypi_0    pypi
rapidfuzz                 3.0.0           py310h7a76584_0    conda-forge
readline                  8.1.2                h3899abd_0    conda-forge
readme-renderer           37.3                     pypi_0    pypi
requests                  2.28.1                   pypi_0    pypi
requests-toolbelt         0.10.1                   pypi_0    pypi
rfc3986                   2.0.0                    pypi_0    pypi
rich                      12.6.0                   pypi_0    pypi
ruamel.yaml               0.18.6          py310hb372a2b_0    conda-forge
ruamel.yaml.clib          0.2.8           py310hb372a2b_0    conda-forge
ruamel.yaml.jinja2        0.2.4                      py_1    conda-forge
scikit-learn              1.2.2           py310h311060b_1    conda-forge
scipy                     1.12.0          py310hdfaad59_2    conda-forge
semver                    3.0.2              pyhd8ed1ab_0    conda-forge
setuptools                65.5.1             pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
snowballstemmer           2.2.0              pyhd8ed1ab_0    conda-forge
soupsieve                 2.5                pyhd8ed1ab_1    conda-forge
sphinx                    7.2.6              pyhd8ed1ab_0    conda-forge
sphinx-book-theme         1.1.0                    pypi_0    pypi
sphinxcontrib-applehelp   1.0.8              pyhd8ed1ab_0    conda-forge
sphinxcontrib-devhelp     1.0.6              pyhd8ed1ab_0    conda-forge
sphinxcontrib-htmlhelp    2.0.5              pyhd8ed1ab_0    conda-forge
sphinxcontrib-jsmath      1.0.1              pyhd8ed1ab_0    conda-forge
sphinxcontrib-qthelp      1.0.7              pyhd8ed1ab_0    conda-forge
sphinxcontrib-serializinghtml 1.1.10             pyhd8ed1ab_0    conda-forge
stack_data                0.6.2              pyhd8ed1ab_0    conda-forge
stdlib-list               0.10.0             pyhd8ed1ab_0    conda-forge
threadpoolctl             3.1.0              pyh8a188c0_0    conda-forge
tk                        8.6.12               h5dbffcc_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
tomli-w                   1.0.0              pyhd8ed1ab_0    conda-forge
tornado                   6.3.2           py310h6729b98_0    conda-forge
traitlets                 5.6.0              pyhd8ed1ab_0    conda-forge
twine                     4.0.2                    pypi_0    pypi
typing_extensions         4.9.0              pyha770c72_0    conda-forge
tzdata                    2022g                h191b570_0    conda-forge
unicodedata2              15.0.0          py310h90acd4f_0    conda-forge
urllib3                   1.26.13                  pypi_0    pypi
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
webencodings              0.5.1                    pypi_0    pypi
wheel                     0.38.4             pyhd8ed1ab_0    conda-forge
xarray                    2022.11.0          pyhd8ed1ab_0    conda-forge
xorg-libxau               1.0.9                h35c211d_0    conda-forge
xorg-libxdmcp             1.1.3                h35c211d_0    conda-forge
xz                        5.2.6                h775f41a_0    conda-forge
zipp                      3.11.0                   pypi_0    pypi
zlib                      1.2.13               hfd90126_4    conda-forge
zstd                      1.5.2                hbc0c0cd_6    conda-forge
```

**Comment by mdhaber** [comment]
Here are my results on Colab with SciPy 1.11.4:

![image](https://github.com/scipy/scipy/assets/6570539/cb79b66f-8fd6-4f76-ba11-1e9f48142861)

Results are similar with 1.12.0.

**Comment by andyfaff** [comment]
I see the same behaviour as the OP on macOS Sonoma, `1.13.0 2.0.0rc1 sys.version_info(major=3, minor=12, micro=2, releaselevel='final', serial=0)`, with the MWE above (adjusted to 1000 iterations) (run via a Jupyter notebook).

![norm_usage_tracking](https://github.com/scipy/scipy/assets/702934/8a2bc2af-01d1-4c1d-9321-752ee703f4f7)

Then I tried moving the random number generation outside the loop:
```
x = np.random.random(size=(100, 50, 50))
mem_vals[0] = process.memory_info().rss / 1e6
for i in range(1, n_elements):
    ppf_val = ndtri(x) # same pattern for norm.ppf
```
and the memory allocation is quite different:
![norm_usage_tracking](https://github.com/scipy/scipy/assets/702934/2fe9df51-2119-4a4e-9d5f-6cc11be91377)

But then I experimented with restarting the kernel and see something akin to the original behaviour. So the first time one runs you get the first curve, the subsequent times you don't see the blowup of memory. 

Quite confusing behaviour, but it's not clear to me that it's a memory leak. It could be some weird interplay of Sonoma memory allocation/Jupyter.

**Comment by djgagne** [comment]
Thank you everyone for testing the code on different systems and configurations. I ran the script myself on a linux system and did not see the memory leak there. I also tried running a bunch of different environments on my Sonoma laptop with numpy and scipy installed from conda, pip, and compiled from source with the latest versions of both. All showed similar growing resource curves, so there must be some kind of weird Sonoma memory allocation thing going on. Given that the issue is not occurring on Linux and there doesn't seem to be anything directly related to the code that is causing the problem, I would be fine closing the issue unless someone else has an idea on how to address this?
